src/hardware/envelope.cpp

```diff
@@ -0,0 +1,111 @@
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+// #define DEBUG 1
+
+#include "support.h"
+#include "envelope.h"
+
+void Envelope::Reactivate()
+{
+	current_edge = 0u;
+	frames_done = 0u;
+	process = &Envelope::Apply;
+}
+
+void Envelope::Update(const uint32_t frame_rate,
+                      const uint32_t peak_magnitude,
+                      const uint8_t expansion_phase_ms,
+                      const uint8_t expire_after_seconds)
+{
+	if (!frame_rate || !peak_magnitude || !expansion_phase_ms)
+		return;
+
+	// How many frames should we inspect before expiring?
+	expire_after_frames = expire_after_seconds * frame_rate;
+	assert(expire_after_frames > 0);
+
+	// The furtherest allowed edge is the peak sample magnitude.
+	edge_limit = peak_magnitude;
+
+	// Permit the envelop to achieve peak volume within the expansion_phase
+	// (in ms) if the samples happen to constantly press on the edges.
+	const uint32_t expansion_phase_frames = ceil_udivide(frame_rate * expansion_phase_ms,
+	                                                     1000u);
+
+	// Calculate how much the envelope's edge will grow after a frame
+	// presses against it.
+	edge_increment = ceil_udivide(peak_magnitude, expansion_phase_frames);
+#ifdef DEBUG
+	LOG_MSG("ENVELOPE: envelope expires after %u frames with a"
+	        "peak permitted magnitude of +/- %u." expire_after_frames,
+	        peak_magnitude);
+#endif
+}
+
+void Envelope::Apply(const bool is_stereo,
+                     const bool is_interpolated,
+                     intptr_t prev[],
+                     intptr_t next[])
+{
+	// Only start the envelope once our samples have actual values
+	if (prev[0] == 0 && frames_done == 0u)
+		return;
+
+	const Bits edge = static_cast<Bits>(current_edge);
+	bool pressed_on_edge = false;
+	if (abs(prev[0]) >= edge) {
+		prev[0] = clamp(prev[0], -edge, edge);
+		pressed_on_edge = true;
+	}
+	if (is_stereo && abs(prev[1]) >= edge) {
+		prev[1] = clamp(prev[1], -edge, edge);
+		pressed_on_edge = true;
+	}
+	if (is_interpolated) {
+		if (abs(next[0]) >= edge) {
+			next[0] = clamp(next[0], -edge, edge);
+			pressed_on_edge = true;
+		}
+		if (is_stereo && abs(next[1]) >= edge) {
+			next[1] = clamp(next[1], -edge, edge);
+			pressed_on_edge = true;
+		}
```

Comment:
@dreamer , this chunk ripe for some nice improvement. 
I've implemented it rote right now, so we can whittle it down perhaps with a lambda, and or using some `min()` and `max()` values.